File: core/engine.py
# The base engine driving Chain Reactions
# Handles all the logic for the game


import queue


# ---------- ON INIT ---------------
SHAPE = None
NTABLE = None


# ----------- INIT -----------------
def init(shape=(9, 6)):
    """ Calculate variables and cache tables """
    global SHAPE, NTABLE

    # store shape
    SHAPE = shape

    # store neighbor indices as tuple of tuples
    s_h, s_w = shape
    NTABLE = [0] * s_w * s_h
    for idx in range(s_h * s_w):
        i_y, i_x = idx // s_w, idx % s_w
        temp = [
            idx - s_w if i_y > 0 else None,
            idx + s_w if i_y < s_h - 1 else None,
            idx - 1 if i_x > 0 else None,
            idx + 1 if i_x < s_w - 1 else None,
        ]
        NTABLE[idx] = tuple([i for i in temp if i is not None])
    NTABLE = tuple(NTABLE)


# ---------- HELPERS ---------------
def territory_count(board, plrid) -> tuple:
    """ Get territory counts (player, enemy) """
    one, two = 0, 0
    for elem in board:
        one += elem < 0
        two += elem > 0
    return (one, two) if plrid else (two, one)


# --------- INTERACTION ------------
def interact_view(board: list, move: int, plrid: int):
    """
    Interact with Chain Reaction Environment
    Returns view of outcome
    """

    board_dupl = board[:]
    gmovr = interact_inplace(board_dupl, move, plrid)

    return (board_dupl, gmovr)


def interact_inplace(board: list, move: int, plrid: int):
    """
    Interact with Chain Reaction Environment
    Modifies board inplace
    """

    # setup
    ntable = NTABLE
    psign = -1 if plrid else 1
    gmovr = False

    # using queue to sequentialize steps
    # near cells are calculated first
    work = queue.Queue()
    work.put(move)

    # store counts (friend, enemy)
    t_frn, t_enm = territory_count(board, plrid)

    while not (work.empty() or gmovr):
        # get next index in queue
        idx = work.get()

        # update territory counts and game over flag
        t_frn += 1
        t_enm -= board[idx] * psign < 0
        gmovr = (t_enm + t_frn > 2) and (t_enm * t_frn == 0)

        # update orb count according to rule
        orbct = abs(board[idx]) + 1
        maxcp = len(ntable[idx])
        board[idx] = (orbct % maxcp) * psign

        # append next indices if exploded
        if orbct == maxcp:
            [work.put(i) for i in ntable[idx]]

    return gmovr


# ---------- GAME CLASS ------------
class GameEngine(object):
    def __init__(self):
        """ Chain Reaction Game Engine """

        # error if not init
        assert SHAPE, "Game Engine not initialized"

        # game state
        self.board = [0] * SHAPE[0] * SHAPE[1]
        self.plrid = 0

        # outcome
        self.gmovr = False
        self.winnr = 2

    def fast_play(self, move) -> bool:
        """
        Calculate the next state of the board
        -------------------------------------
        Input   : move index (tuple or int)
        Returns : success boolean
        """

        # setup
        idx = move[0] * SHAPE[1] + move[1] if type(move) == tuple else move
        psign = -1 if self.plrid else 1

        # invalid condition
        if (self.board[idx] * psign < 0) or self.gmovr:
            return False

        # interact inplace
        self.gmovr = interact_inplace(self.board, idx, self.plrid)
        self.winnr = self.plrid if self.gmovr else 2

        # toggle player
        self.plrid = 1 - self.plrid
        return True

File: csource/src/engine.c
#include "chain/cqueue.h"
#include "chain/engine.h"


/* Critical Mass Lookup Table */
static const int NTABLE [9 * 6] = {
    2, 3, 3, 3, 3, 2,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    3, 4, 4, 4, 4, 3,
    2, 3, 3, 3, 3, 2,
};


/* Interact with environment */
int
engine__interact ( int  *old_board,
                   int  *new_board,
                   int   move,
                   int   player )
{
    int game_over = 0;
    int psign     = player ? -1 : 1;
    int t_frn     = 0;
    int t_enm     = 0;

    /* queue acts as reactions sequence */
    CQueue *work = cqueue__create(100);
    cqueue__enqueue(work, move);

    /* count territories while copying whole board */
    for (int i = 0; i < 54; ++i)
    {
        new_board[i] = old_board[i];         // copy
        if      (new_board[i] > 0)  ++t_frn; // positive
        else if (new_board[i] < 0)  ++t_enm; // negative
    }

    /* swap enemy and friend counts if player is 1 */
    if (player)
    {
        int temp = t_frn;
        t_frn = t_enm;
        t_enm = temp;
    }

    /* stop if game over or the queue is empty */
    while ((!game_over) && (!cqueue__isempty(work)))
    {
        /* get next index in queue */
        move = cqueue__dequeue(work);

        /* update territory counts and game over */
        t_frn += 1;
        t_enm -= (new_board[move] * psign < 0);
        game_over = ((t_enm + t_frn > 2) && (t_enm * t_frn == 0));
    
        /* update orb count */
        int cell = new_board[move];
        int absl = (cell > 0) ? cell : -cell;
        new_board[move] = ((absl + 1) % NTABLE[move]) * psign;

        /* add neighbors to queue if exploded */
        if (absl + 1 == NTABLE[move])
        {
            int i_y = move / 6;
            int i_x = move % 6;

            if (i_y > 0)  { cqueue__enqueue(work, move - 6); }
            if (i_y < 8)  { cqueue__enqueue(work, move + 6); }
            if (i_x > 0)  { cqueue__enqueue(work, move - 1); }
            if (i_x < 5)  { cqueue__enqueue(work, move + 1); }
        }    
    }

    /* cleanup */
    cqueue__destory(work);
    return game_over;
}
